brainiak/isfc.py

```diff
@@ -1,226 +1,1104 @@
+        
+    summary_statistic : numpy function, default:np.median
+        Summary statistic, either np.median (default) or np.mean
+
+    n_permutations : int, default:1000
+        Number of permutation iteration (randomizing group assignment)
+        
+    return_distribution : bool, default:False
+        Optionally return the bootstrap distribution of summary statistics
+        
+    random_state = int, None, or np.random.RandomState, default:None
+        Initial random seed
+
+    Returns
+    -------
+    observed : float, ISC summary statistic or difference
+        Actual ISC or group difference (excluding between-group ISCs)
+
+    p : float, p-value
+        p-value based on permutation test
+        
+    distribution : ndarray, permutations by voxels (optional)
+        Permutation distribution if return_bootstrap=True
+    """
+    
+    # Standardize structure of input data
+    if type(iscs) == list:
+        iscs = np.array(iscs)[:, np.newaxis]
+        
+    elif type(iscs) == np.ndarray:
+        if iscs.ndim == 1:
+            iscs = iscs[:, np.newaxis]
+
+    # Check if incoming pairwise matrix is vectorized triangle
+    if pairwise:
+        try:
+            test_square = squareform(iscs[:, 0])
+            n_subjects = test_square.shape[0]
+        except ValueError:
+            raise ValueError("For pairwise input, ISCs must be the "
+                             "vectorized triangle of a square matrix.")
+    elif not pairwise:
+        n_subjects = iscs.shape[0]
+            
+    # Check match between group labels and ISCs
+    if type(group_assignment) == list:
+        pass
+    elif type(group_assignment) == np.ndarray:
+        group_assignment = group_assignment.tolist()
+    else:
+        print("No group assignment provided, performing one-sample test.")
+    
+    if group_assignment and len(group_assignment) != n_subjects:
+        raise ValueError(f"Group assignments ({len(group_assignment)}) "
+                         f"do not match number of subjects ({n_subjects})!")
+    
+    # Set up group selectors for two-group scenario
+    if group_assignment and len(np.unique(group_assignment)) == 2:
+        n_groups = 2
+        
+        # Get group labels and counts
+        group_labels = np.unique(group_assignment)
+        groups = {group_labels[0]: group_assignment.count(group_labels[0]),
+                  group_labels[1]: group_assignment.count(group_labels[1])}
+
+        # For two-sample pairwise approach set up selector from matrix
+        if pairwise == True:
+            # Sort the group_assignment variable if it came in shuffled
+            # so it's easier to build group assignment matrix
+            sorter = np.array(group_assignment).argsort()
+            unsorter = np.array(group_assignment).argsort().argsort()
+            
+            # Populate a matrix with group assignments                                 
```

Comment:
This line is hard to read/understand. 
Should we add an intermediate step (e.g. by giving `np.full((groups[group_labels[0]], groups[group_labels[1]]), group_labels[0])))` and `np.full((groups[group_labels[0]], groups[group_labels[1]]), np.nan))` some varaible names?